src/java/org/broadinstitute/dropseqrna/metrics/CountUnmatchedSampleIndices.java

```diff
@@ -0,0 +1,191 @@
+        final TreeSet<IndexAndCount> indexAndCounts = new TreeSet<>();
+
+        // Note that there for sample indices with counts that are equal, at the tail end of MAX_OUTPUT, it is
+        // arbitrary which will be included in the output.
+        for (Map.Entry<String, LongAdder> entry: tally.entrySet()) {
+            final int count = entry.getValue().intValue();
+            if (indexAndCounts.size() < MAX_OUTPUT) {
+                indexAndCounts.add(new IndexAndCount(entry.getKey(), count));
+            } else if (count > indexAndCounts.first().count) {
+                indexAndCounts.remove(indexAndCounts.first());
+                indexAndCounts.add(new IndexAndCount(entry.getKey(), count));
+            }
+        }
+
+        // Create metrics beans for the most frequent sample indices, and write to file.
+        MetricsFile<UnmatchedSampleIndexMetrics, Integer> outFile = getMetricsFile();
+        for (final IndexAndCount indexAndCount: indexAndCounts.descendingSet()) {
+            outFile.addMetric(new UnmatchedSampleIndexMetrics(indexAndCount.index, indexAndCount.count,
+                    indexAndCount.count/(double)totalUnmatchedReads,
+                    indexAndCount.count/(double)totalReads));
+        }
+
+        outFile.write(OUTPUT);
+
+        return 0;
+    }
+
+    private class Worker
+            implements Callable<Boolean> {
+        private final Iterator<File> barcodeFiles;
+        private final ConcurrentHashMap<String, LongAdder> map;
+        int totalReads = 0;
+        int totalUnmatchedReads = 0;
+
+        public Worker(Iterator<File> barcodeFiles, ConcurrentHashMap<String, LongAdder> map) {
+            this.barcodeFiles = barcodeFiles;
+            this.map = map;
+        }
+
+        @Override
+        public Boolean call() throws Exception {
+            for (final File barcodeFile: new IterableAdapter<>(barcodeFiles)) {
+                LOG.debug("Processing", barcodeFile);
+                final TabbedInputParser parser = new TabbedInputParser(false, barcodeFile);
+                for (final String[] row : parser) {
+                    ++totalReads;
+                    if ("N".equals(row[1])) {
+                        ++totalUnmatchedReads;
+                        map.computeIfAbsent(row[0], k -> new LongAdder()).increment();
+                    }
+                }
+                CloserUtil.close(parser);
+            }
+            return true;
+        }
+    }
+
+    private static class IndexAndCount
+            implements Comparable<IndexAndCount> {
+        final String index;
+        final int count;
+
+        public IndexAndCount(String index, int count) {
+            this.index = index;
+            this.count = count;
+        }
+
+        @Override
+        public int compareTo(IndexAndCount o) {
+            return Integer.compare(this.count, o.count);
+        }
+    }
+
+    // I can't quite believe there isn't something that already exists in java.concurrent
```

Comment:
Hi @jamesnemesh ,

This is not particularly important, so just ignore if you don't have time.  I can't quite believe that there isn't an existing Java class that provides an iterator that can be called by multiple threads and guarantees that elements will be returned once and only once, but the various classes I tried (CopyOnWriteArrayList, ArrayBlockingQueue) didn't appear to have this property.  Do you have any ideas?

Thanks, Alec